probe/plugins/registry.go

```diff
@@ -0,0 +1,308 @@
+func NewPlugin(ctx context.Context, socket string, client *http.Client, expectedAPIVersion string, handshakeMetadata map[string]string) *Plugin {
+	params := url.Values{}
+	for k, v := range handshakeMetadata {
+		params.Add(k, v)
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	p := &Plugin{context: ctx, socket: socket, client: client, cancel: cancel}
+	f := p.handshake(ctx, expectedAPIVersion, params)
+	// try the first time synchronously (makes testing easier)
+	if ok, err := f(); !ok {
+		log.Errorf("Error plugin handshake, backing off 10s: %v", err)
+		p.backoff = backoff.New(f, "plugin handshake")
+		go func() {
+			time.Sleep(10 * time.Second)
+			p.backoff.Start()
+		}()
+	}
+	return p
+}
+
+type handshakeResponse struct {
+	Name        string   `json:"name"`
+	Description string   `json:"description,omitempty"`
+	Interfaces  []string `json:"interfaces"`
+	APIVersion  string   `json:"api_version,omitempty"`
+}
+
+// handshake tries the handshake with this plugin.
+func (p *Plugin) handshake(ctx context.Context, expectedAPIVersion string, params url.Values) func() (bool, error) {
+	return func() (ok bool, err error) {
+		defer func() { p.setStatus(err) }()
+		var resp handshakeResponse
+		if err := p.get("/", params, &resp); err != nil {
+			return err == context.Canceled, fmt.Errorf("plugins: error loading plugin %s: %v", p.socket, err)
+		}
+
+		if resp.Name == "" {
+			return false, fmt.Errorf("plugins: error loading plugin %s: plugin did not provide a name", p.socket)
+		}
+		if resp.APIVersion != expectedAPIVersion {
+			return false, fmt.Errorf("plugins: error loading plugin %s: plugin did not provide correct API version: expected %q, got %q", p.socket, expectedAPIVersion, resp.APIVersion)
+		}
+		p.ID, p.Label = resp.Name, resp.Name
+		p.Description = resp.Description
+		p.Interfaces = resp.Interfaces
+		log.Infof("plugins: loaded plugin %s: %s", p.ID, strings.Join(p.Interfaces, ", "))
+		return true, nil
+	}
+}
+
+// Report gets the latest report from the plugin
+func (p *Plugin) Report() (report.Report, error) {
+	result := report.MakeReport()
+	err := p.get("/report", nil, &result)
+	p.setStatus(err)
+	return result, err
+}
+
+func (p *Plugin) setStatus(err error) {
+	if err == nil {
+		p.Status = "ok"
+	} else {
+		p.Status = fmt.Sprintf("error: %v", err)
+	}
+}
+
+// TODO(paulbellamy): better error handling on wrong status codes
+func (p *Plugin) get(path string, params url.Values, result interface{}) error {
+	// Context here lets us either timeout req. or cancel it in Plugin.Close
+	ctx, cancel := context.WithTimeout(p.context, pluginTimeout)
+	defer cancel()
+	resp, err := ctxhttp.Get(ctx, p.client, fmt.Sprintf("unix://%s?%s", path, params.Encode()))
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("plugin returned non-200 status code: %s", resp.Status)
+	}
+	defer resp.Body.Close()
+	if err := codec.NewDecoder(resp.Body, &codec.JsonHandle{}).Decode(&result); err != nil {
```

Comment:
We should probably perform a maximum  `Content-length` check (something between 50-100MB?) before blinding parsing the response to prevent plugins from unintentionally DoS'ing the probe.